src/main/java/com/epam/javacourses2016/task19/SolverTask19.java

```diff
@@ -17,8 +19,93 @@
+        int count = 0;
+        for (Car firstCar : cars) {
+            for (Car secondCar : cars) {
+                count += getOvertake(firstCar, secondCar, lengthLap, numberLaps);
+            }
+        }
+        return count;
+    }
+
+    private int getOvertake(Car firstCar, Car secondCar, long lengthLap, int numberLaps) {
+
+        double numberOfOvertakes = 0;
+        if (firstCar.getSpeed() <= secondCar.getSpeed()) {
+            return (int) numberOfOvertakes;
+        }
+        double lengthFirstCar = lengthLap * numberLaps - firstCar.getStartPosition();
+        double timeFirstCar = lengthFirstCar / firstCar.getSpeed();
+
+        double lengthSecondCar = timeFirstCar * secondCar.getSpeed() + secondCar.getStartPosition();
+        double lapsSecondCar = lengthSecondCar / lengthLap;
+
+        numberOfOvertakes = numberLaps - lapsSecondCar;
+
+        if (firstCar.getStartPosition() < secondCar.getStartPosition()) {
+            numberOfOvertakes++;
+        }
+        return (int) numberOfOvertakes;
+    }
+
+    /**
+     * @param cars       Расположенные на трассе машины.
+     * @param lengthLap  Длина трассы.
+     * @param numberLaps Количество кругов.
+     * @return Количество осуществленных обгонов.
+     */
     int getNumberOvertaking(Set<Car> cars, long lengthLap, int numberLaps) {
-        //TODO
-        return 0;
+        Set<Car> sortedCars = new TreeSet<>(new CarCompatator());
+        sortedCars.addAll(cars);
+        int numberOfOvertakes = 0;
+        while (!raceComplete(sortedCars, numberLaps)) {
+            int i = 0;
+            int timeDelta = 1;
+            for (Car car : sortedCars) {
+                car.setActualPosition(i);
+                i++;
+                int newDistance = timeDelta * car.getSpeed();
+                if (car.getDistanceCovered() + newDistance <= lengthLap) {
+                    car.setDistanceCovered(car.getDistanceCovered() + newDistance);
+                } else {
+                    car.setDistanceCovered(car.getDistanceCovered() + newDistance);
+                    car.setNumberOfFinishedLaps(car.getNumberOfFinishedLaps() + 1);
+                }
+            }
+            Set<Car> sortedCarsSecond = new TreeSet<>(new CarCompatator());
+            sortedCarsSecond.addAll(sortedCars);
+            sortedCars.clear();
+            i = 0;
+            for (Car car : sortedCarsSecond) {
+                if (i > car.getActualPosition()) {
+                    numberOfOvertakes = numberOfOvertakes + (i - car.getActualPosition());
+                }
+                i++;
+                if (car.getNumberOfFinishedLaps() < numberLaps) {
+                    if (car.getDistanceCovered() >= lengthLap) {
+                        car.setDistanceCovered(car.getDistanceCovered() - lengthLap);
+                    }
+                    sortedCars.add(car);
+                }
+            }
+        }
+        return numberOfOvertakes;
```

Comment:
You should think about decomposition such a big methods with several loops and if-sections